src/main/java/se/liu/ida/hefquin/queryplan/executable/impl/ops/ExecOpSymmetricHashJoin.java

```diff
@@ -0,0 +1,91 @@
+package se.liu.ida.hefquin.queryplan.executable.impl.ops;
+
+import org.apache.jena.graph.Node;
+import org.apache.jena.sparql.core.Var;
+import org.apache.jena.sparql.engine.binding.Binding;
+import se.liu.ida.hefquin.data.SolutionMapping;
+import se.liu.ida.hefquin.data.jenaimpl.JenaBasedSolutionMapping;
+import se.liu.ida.hefquin.data.jenaimpl.JenaBasedSolutionMappingUtils;
+import se.liu.ida.hefquin.queryplan.executable.IntermediateResultBlock;
+import se.liu.ida.hefquin.queryplan.executable.IntermediateResultElementSink;
+import se.liu.ida.hefquin.queryproc.ExecutionContext;
+
+import java.util.*;
+
+public class ExecOpSymmetricHashJoin extends ExecOpGenericJoin{
+    protected final Map<ListKey<Node>, JenaBasedSolutionMapping> mapL = new HashMap<>();
+    protected final Map<ListKey<Node>, JenaBasedSolutionMapping> mapR = new HashMap<>();
+
+    @Override
+    public int preferredInputBlockSize() {
+        return 1;
+    }
+
+    @Override
+    public boolean requiresCompleteChild1InputFirst() {
+        return false;
+    }
+
+    @Override
+    public void processBlockFromChild1(IntermediateResultBlock input,
+                                       IntermediateResultElementSink sink,
+                                       ExecutionContext execCxt) {
+        final List<SolutionMapping> liL = input.ArrayList();
+        for(SolutionMapping SolMapL: liL){
+            final JenaBasedSolutionMapping JBSolMapL = (JenaBasedSolutionMapping) SolMapL;
+            final Binding LHS = JBSolMapL.asJenaBinding();
+
+            // composite key
+            final ArrayList<Node> valKeyL = new ArrayList<>();
+            for (Var v : keys) {
+                valKeyL.add(LHS.get(v));
+            }
+            // Insert into left hash map
+            if (! valKeyL.isEmpty()){
+                mapL.put(new ListKey(valKeyL), JBSolMapL);
```

Comment:
This replaces the previous entry at the given hash key (if there was any). That must not happen!

Note, multiple solution mappings may have the same hash key (because they have the same values for the join variables). They must all be kept in the hash table!